% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/coverage.plot.R
\name{coverage.plot}
\alias{coverage.plot}
\title{Plot the bin coverage in a region}
\usage{
coverage.plot(chr, start, end, bc.f, norm.stats.f = NULL, sv.df = NULL,
  ref.samples = NULL, boxplot = FALSE, samples = NULL, files.df = NULL,
  anno.df = NULL, anno.col = "geneName", flanks = 10000,
  absolute.position = TRUE, nb.cores = 1)
}
\arguments{
\item{chr}{the chromosome}

\item{start}{the start position of the region}

\item{end}{the end position of the region}

\item{bc.f}{the path to the bin count file.}

\item{norm.stats.f}{the path to the normalization statistics file.}

\item{sv.df}{the data.frame with the calls.}

\item{ref.samples}{the names of the reference samples.}

\item{boxplot}{should the reference be represented as boxplots. If FALSE, violin plots will be used.}

\item{samples}{the set of samples to represent.}

\item{files.df}{a data.frame with the path to the different files associated with each sample. If 'sample' is absent from 'bc.f' the correct file from 'files.df' will be used. Default is NULL.}

\item{anno.df}{a data.frame with additional information (e.g. gene annotation) to be added to the graph.}

\item{anno.col}{the name of 'anno.df' column to use to differentiate elements in the graph. E.g. 'geneName' to color the genes in the graph.}

\item{flanks}{the size of the flanking region. Default is 10000.}

\item{absolute.position}{should the bins be placed in absolute position. Default is TRUE. If FALSE, distant bins might be displayed next to each other if no bins are available in between (useful for targeted sequencing).}

\item{nb.cores}{the number of cores to use. Default is 1.}
}
\value{
a ggplot object
}
\description{
Plot the bin coverage in a genomic regions across samples. The files ('bc.f', 'norm.stats.f') must be ordered and indexed.
}
\author{
Jean Monlong
}
